README.rst
========
bibclean
========


CHANGES.rst
v0.0.1, 2014-03-16 -- Initial Release


bibclean/tests/test_references.py
import references


bibclean/tests/test_peoplenames.py
import unittest
import peoplenames


class TestUnicodeFunctions(unittest.TestCase):
    def test_name_to_ascii(self):
        name = ('Ó Súilleabháin', 'Jörg Tanjō')
        rnme = peoplenames.name_to_ascii(name)
        tnme = ("O'Suilleabhain", 'Jorg Tanjo')
        self.assertEqual(rnme, tnme)


class TestListFunctions(unittest.TestCase):
    def test_break_name(self):
        names = [('(George) Ó Súilleabháin', 'Jörg-T.'),
                 ('Smith', 'Jane Kate'),
                 ("O'Brian", 'James T.')]

        tnmes = [[[('George', 'G'), ('Ó Súilleabháin', 'Ó')],
                  [('Jörg', 'J'), ('T', 'T')]],
                 [[('Smith', 'S')], [('Jane', 'J'), ('Kate', 'K')]],
                 [[("O'Brian", 'O')], [('James', 'J'), ('T', 'T')]]]
        for idx, name in enumerate(names):
            self.assertEqual(peoplenames.break_name(name), tnmes[idx])

    def test_break_name_simple(self):
        name = ('Smith', 'Jane Kate')
        rnme = peoplenames.break_name(name)
        tnme = [[('Smith', 'S')],
                [('Jane', 'J'), ('Kate', 'K')]]
        self.assertEqual(rnme, tnme)

    def test_break_name_complex(self):
        name = ("O'Brian", 'James T.')
        rnme = peoplenames.break_name(name)
        tnme = [[("O'Brian", 'O')],
                [('James', 'J'), ('T', 'T')]]
        self.assertEqual(rnme, tnme)

    def test_part_comp(self):
        parts = [('Jane', 'J'), ('Kate', 'K')]
        tprts = [['Jane', 'Kate'], ['J', 'K']]
        for idx, tprt in enumerate(tprts):
            self.assertEqual(tprt, peoplenames.part_comp(parts, idx))

    def test_inv_idx(self):
        self.assertEqual([0, 2, 3], peoplenames.inv_idx([1, 4], 5))


class TestNameComparisons(unittest.TestCase):
    def test_name_comp(self):
        namesets = [
            (2+1+1/4, ('Smith', 'Jane Kate'), ('Smith', 'Jane K.')),
            (1.75+1.75/2+1/4, ('Ó Súilleabháin', 'Jörg Tanjō'),
             ("O'Suilleabhain", 'Jorg T.'))]
        for ns in namesets:
            self.assertEqual(ns[0], peoplenames.name_comp(ns[1], ns[2]))

if __name__ == '__main__':
    unittest.main()


cleanZot.py
#!/usr/bin/env python

from unidecode import unidecode
from pyzotero import zotero
import re
from datetime import datetime
from dateutil import parser as dup
import argparse

parser = argparse.ArgumentParser()
parser.add_argument("library_id", help="Zotero API user ID", type=str)
parser.add_argument("api_key", help="Zotero API user key", type=str)
args = parser.parse_args()

zot = zotero.Zotero(args.library_id, 'user', args.api_key)

itms = zot.everything(zot.top())

auths = {}
itm_key = {}
itm_auths = {}
odd_names = []
odd_fnames_idx = {}
odd_dates_idx = {}

reOLN = re.compile(r'[^a-zA-Z]')
reOFN = re.compile(r'(^[a-z])|([^a-zA-Z -\.])')
reODT = re.compile(r'^\d{4}(?:-\d{2}){0,2}$')
reFDY = re.compile(r'\b(\d{1,2})\b')
for idx, itm in enumerate(itms):
    key = itm['key']
    itm_key[key] = idx
    if 'date' in itm:
        if not reODT.search(itm['date']) and itm['date'] != '':
            dts = itm['date']
            dto = dup.parse(dts)
            dys = [int(d) for d in reFDY.findall(dts)]
            odd_dates_idx[key] = (dto, dts, dys)
    if 'creators' in itm:
        itm_auths[itm['key']] = []
        for crtr in itm['creators']:
            if 'firstName' in crtr and 'lastName' in crtr:
                name_last = unidecode(crtr['lastName'])
                name_first = unidecode(crtr['firstName'])
                if not name_last in auths:
                    auths[name_last] = {}
                    if reOLN.search(name_last):
                        odd_names.append(name_last)
                if not name_first in auths[name_last]:
                    auths[name_last][name_first] = []
                if reOFN.search(name_first):
                    if not name_first in odd_fnames_idx:
                        odd_fnames_idx[name_first] = set()
                    odd_fnames_idx[name_first].add(key)
                auths[name_last][name_first].append(key)
                itm_auths[key].append((name_last, name_first))
odd_names.sort()
odd_fnames = list(odd_fnames_idx.keys())
odd_fnames.sort()

auth_counts = {}
for lastn, tmp in auths.items():
    auth_counts[lastn] = len(tmp)
auth_order = sorted(auth_counts, key=auth_counts.get, reverse=True)


def print_items(kys):
    for itm in kys:
        idx = itm_key[itm]
        print(itms[idx]['date'],
              itms[idx]['publicationTitle'],
              itms[idx]['title'])


def has_initials(name, initials):
    name_inits = re.findall('([A-Z])', name)
    for init in name_inits:
        if init in initials:
            return True
    return False


match_name = re.compile('\(?(([A-Z])[a-zA-Z]*)\)?')
name_inits = re.compile('([A-Z])')
for nm in odd_names:
    ms = match_name.findall(nm)
    ofi = []
    for ofnm in list(auths[nm].keys()):
        ofi += name_inits.findall(ofnm)
    for idm, m in enumerate(ms):
        if m[0] in auths:
            oms = [seq[1] for seq in (ms[:idm] + ms[(idm+1):])] + ofi
            for fnm in list(auths[m[0]].keys()):
                if has_initials(fnm, oms):
                    print(nm, '; ', ', '.join(list(auths[nm].keys())),
                          ': ', m[0], ', ', fnm, sep='')


def name_comp_score(name1, name2):
    score = 0.0
    if name1[0] != name2[0]:
        return 0
    brkname = re.compile(r'(([A-Z])[a-z]*)')
    fna1 = brkname.findall(name1[1])
    fna2 = brkname.findall(name2[1])
    if len(fna2) < len(fna1):
        tmp = fna1
        fna1 = fna2
        fna2 = tmp

    for idx, fnm in enumerate(fna1):
        if fnm[0] == fna2[idx][0]:
            score += 1/(idx+1)
        elif fnm[1] == fna2[idx][1]:
            score += 0.5/(idx+1)
    return score


def name_list_scores(names):
    name_scores = []
    for i in range(0, len(names)-2):
        score = name_comp_score(names[i], names[i+1])
        if score > 0:
            name_scores.append((score, names[i], names[i+1]))
    return name_scores


def name_occrs(names):
    if len(names) < 1:
        return []
    tstname = names[0]
    count = 1
    i = 1
    res = []
    while (i < len(names)):
        score = name_comp_score(tstname, names[i])
        if score < 0.5:
            res.append((count, tstname))
            tstname = names[i]
            count = 0
        elif len(tstname[1]) < len(names[i][1]):
            tstname = names[i]
        count += 1
        i += 1
    res.append((count, tstname))
    return res


def name_list_comp(names1, names2):
    comp_list = []
    nmslst = {}
    for nm1 in names1:
        if not nm1[1][0] in nmslst:
            nmslst[nm1[1][0]] = []
        nmslst[nm1[1][0]].append(nm1)
    for nm2 in names2:
        if nm2[1][0] in nmslst:
            for nm1 in nmslst[nm2[1][0]]:
                score = name_comp_score(nm1[1], nm2[1])
                if score >= 0.25:
                    cmbscore = score*(nm1[0]+nm2[0])
                    comp_list.append((cmbscore, score, nm1, nm2))
    return comp_list


def list_coauths(keys, ign_list):
    coauths = []
    for key in keys:
        coauths += [itm for itm in itm_auths[key] if itm not in ign_list]
    coauths.sort()
    return coauths


def comp_fnames(lname, fname1, fname2):
    ca1 = name_occrs(list_coauths(auths[lname][fname1], [(lname, fname1)]))
    ca2 = name_occrs(list_coauths(auths[lname][fname2], [(lname, fname2)]))
    return name_list_comp(ca1, ca2)


def fix_dates():
    for key, val in odd_dates_idx.items():
        if val[0].day != val[0].month and val[0].day in val[2]:
            itms[itm_key[key]]['date'] = val[0].strftime('%Y-%m-%d')
            if zot.check_items([itms[itm_key[key]]]):
                itms[itm_key[key]] = zot.update_item(itms[itm_key[key]])
                print(itms[itm_key[key]]['key'], key,
                      itms[itm_key[key]]['date'], val[0].strftime('%Y-%m-%d'))


prox = 'http://ezproxy.rit.edu/login?url=$@'


def fix_proxy(proxy):
    prox_brk = re.search(r'^(?:https?://)([^/\?]+)(.*)\$\@$', prox)
    if prox_brk:
        rePRB = re.compile(r'^(https?://)?(?:(.+)\.|)' +
                           re.escape(prox_brk.group(1)) +
                           r'(?:' +
                           re.escape(prox_brk.group(2)) +
                           r'https?://)?(.*)$')
        rePRD = re.compile(re.escape(prox_brk.group(1)))
        for idx, itm in enumerate(itms):
            url = itm['url']
            key = itm['key']
            if rePRD.search(url):
                nurl = ''.join([m
                                for m in rePRB.search(url).groups()
                                if m is not None])
                itms[idx]['url'] = nurl
                if zot.check_items([itms[idx]]):
                    itms[idx] = zot.update_item(itms[idx])


def brk_name(name):
    parts = []
    for nm in re.findall(r'\b(([a-z])[a-z]*)\b|(([A-Z])[a-z]*)', name):
        if nm[0] == '':
            parts.append(nm[2:])
        else:
            parts.append(nm[:2])
    return parts


def name_rpts(parts):
    inits = ''.join([p[1].capitalize() for p in parts])
    norep_inits = re.search(r'^(.*?)\1*$', inits).group(1)
    if norep_inits != inits:
        nms = grp_names([p[0].capitalize() for p in parts], len(norep_inits))
        new_full = [beg_str_equal(nm) for nm in nms]
        if None not in new_full:
            brkname = re.compile(r'(([A-Z])[a-z]*)')
            parts = []
            for nm in new_full:
                parts += brkname.findall(nm)
    return parts


def beg_str_equal(strs):
    strs.sort(key=len, reverse=True)
    for k in range(1, len(strs)):
        if strs[k-1][:len(strs[k])] != strs[k]:
            return None
    return strs[0]


def grp_names(names, n):
    rtn = []
    for k in range(0, n):
        rtn.append([names[i] for i in range(k, len(names), n)])
    return rtn


def fnd_rep(names, thresh):
    rtn = []
    for idx, name in enumerate(names):
        parts = brk_name(name)
        if len(parts) >= thresh:
            rmr_parts = name_rpts(parts)
            if len(parts) != len(rmr_parts) and len(rmr_parts) >= thresh:
                rtn.append((idx, name, parts, rmr_parts))
    return rtn


try:
    zot = zotero.Zotero('', 'user', '')
except:
    print('Could not connect')


class Person(object):
    """
    Author Info
    """
    def __init__(self, name_first, name_last):
        self._name = [name_last, name_first]
        self._ascii_name = [re.sub(r'`', '\'', unidecode(n))
                            for n in self._name]

    def name(self, idx, uni):
        if uni:
            return self._name[idx]
        else:
            return self._ascii_name[idx]

    def name_last(self, *args, **kwargs):
        uni = kwargs.get('uni', False)
        return self.name(0, uni)

    def name_first(self, *args, **kwargs):
        uni = kwargs.get('uni', True)
        return self.name(1, uni)

    def update(self):
        pass

ppl = []
for x in itms[239]['creators']:
    if 'firstName' in x and 'lastName' in x:
        ppl.append(Person(x['firstName'], x['lastName']))

for idi, itm in enumerate(ref.items):
    if itm is None:
        print('Item:', str(idi))
    else:
        for idc, con in enumerate(itm.contributors):
            if con is None:
                print('Item', str(idi), 'Con:', str(idc))

for ln, fns in ref.people.items():
    if fns is None:
        print('No FNS:', ln)
    else:
        for fn, per in fns.items():
            if per is None:
                print('No Person:', ln, fn)
            else:
                for idx, con in enumerate(per.contributions):
                    if con is None:
                        print('No Contrib:', ln, fn, '; id:', idx)


bibclean/clean.py
import regex
import references
import json
import urllib.request
from dateutil import parser


def frequent_person(people):
    pass


def comp_people(people):
    pass


def remove_ezproxies(items, *args, **kwargs):
    proxy_list_url = kwargs.get('proxy_list_url',
                                'http://ezproxy-db.appspot.com/proxies.json')
    proxy_list = kwargs.get('proxy_list', None)
    proxy_names = kwargs.get('proxy_name', None)
    proxy = kwargs.get('proxy', None)
    if proxy is not None:
        return remove_ezproxy(items, proxy)
    if proxy_list is None:
        proxy_list = {}
        site = urllib.request.urlopen(proxy_list_url)
        proxies = json.loads(site.read().decode())
        for p in proxies:
            proxy_list[p['name']] = p['url']
    if proxy_names is None:
        proxy_names = list(proxy_list.keys())
    for pn in proxy_names:
        if pn in proxies:
            remove_ezproxy(items, proxy_list[pn])


def remove_ezproxy(items, proxy):
    prox_brk = regex.search(r'^(?:https?://)([^/\?]+)(.*)\$\@$', proxy)
    if prox_brk:
        rePRB = regex.compile(r'^(https?://)?(?:(.+)\.|)' +
                           re.escape(prox_brk.group(1)) +
                           r'(?:' +
                           re.escape(prox_brk.group(2)) +
                           r'https?://)?(.*)$')
        rePRD = regex.compile(re.escape(prox_brk.group(1)))
        for item in items:
            if rePRD.search(item.url):
                item.url = ''.join([m
                                    for m in rePRB.search(item.url).groups()
                                    if m is not None])


def fix_dates(items):
    reODT = regex.compile(r'^\d{4}(?:-\d{2}){0,2}$')
    reFDY = regex.compile(r'\b(\d{1,2})\b')
    needs_fix = []
    for item in items:
        if not reODT.search(item.date) and item.date != '':
            date_obj = parser.parse(item.date)
            date_days = [int(d) for d in reFDY.findall(item.date)]
            if date_obj.day != date_obj.month and date_obj.day in date_days:
                item.date = date_obj.strftime('%Y-%m-%d')
            else:
                needs_fix.append((item, date_obj, date_days))


bibclean/references.py
from pyzotero import zotero
from peoplenames import name_comp, name_to_ascii
import weakref


class References(object):
    """
    Parent of reference sources
    """
    def __init__(self, dbc, dbc_params, *cargs, **kwargs):
        self.test = kwargs.get('test', False)
        self.dbc = zotero.Zotero(dbc_params['library_id'],
                                 'user', dbc_params['api_key'])
        self.dbc_params = dbc_params
        self.items = []
        self.people = {}

    def get_all(self):
        items_raw = self.dbc.everything(self.dbc.top())
        for item_raw in items_raw:
            self.add_item(item_raw)

    def add_item(self, raw):
        self.items.append(Item(self, raw))

    def get_person(self, name):
        if not name[0] in self.people:
            self.people[name[0]] = {}
        if name[1] not in self.people[name[0]]:
            self.people[name[0]][name[1]] = Person(name)
        return self.people[name[0]][name[1]]


class Item(object):
    """
    Item
    """
    def __init__(self, parent, raw):
        self._raw = raw
        self.parent = weakref.proxy(parent)
        self.contributors = []
        self._add_contributor()

    def update(self):
        if self.parent.test:
            return True
        if self.parent.dbc.check_items([self._raw]):
            try:
                self._raw = self.parent.update_item(self._raw)
                return True
            except:
                raise
        return False

    def _update_field(self, field, value):
        tmp = self._raw[field] = value
        self._raw[field] = value
        try:
            self.update()
        except:
            self._raw[field] = tmp
            raise

    def _field_value(self, field):
        if field in self._raw:
            return self._raw[field]
        else:
            return None

    def _add_contributor(self):
        if 'creators' in self._raw:
            for creator in self._raw['creators']:
                if 'firstName' in creator and 'lastName' in creator:
                    name = (creator['lastName'], creator['firstName'])
                    person = self.parent.get_person(name)
                    contrib = Contributor(self, person, name)
                    self.contributors.append(contrib)
                    person.contributions.add(contrib)

    @property
    def title(self):
        return self._field_value('title')

    @title.setter
    def title(self, value):
        return self._update_field('title', value)

    @property
    def title_short(self):
        return self._field_value('shortTitle')

    @title.setter
    def title(self, value):
        return self._update_field('shortTitle', value)

    @property
    def url(self):
        return self._field_value('url')

    @url.setter
    def url(self, value):
        self._update_field('url', value)

    @property
    def journal(self):
        return self._field_value('publicationTitle')

    @journal.setter
    def journal(self, value):
        self._update_field('publicationTitle', value)

    @property
    def journal_abbrev(self):
        return self._field_value('journalAbbreviation')

    @journal_abbrev.setter
    def journal_abbrev(self, value):
        self._update_field('journalAbbreviation', value)

    @property
    def date(self):
        self._field_value('date')

    @date.setter
    def date(self, value):
        self._update_field('date', value)

    def update_contributor(self, idx, name):
        if 'creators' in self._raw:
            self._raw['creators'][idx]['lastName'] = name[0]
            self._raw['creators'][idx]['firstName'] = name[1]
            self.update()


class Person(object):
    """
    Person
    """
    def __init__(self, name):
        self.contributions = weakref.WeakSet()
        self._name = name

    @property
    def ascii_name(self):
        return name_to_ascii(self._name)

    @property
    def name(self):
        return self._name

    @name.setter
    def name(self, value):
        for c in self.contributions:
            if c.name != value:
                c.name = value
        self._name = value

    def compare(self, person):
        return name_comp(self.name, person.name)

    def cocontributors(self):
        cocontrib = []
        counts = []
        for contrib in self.contributions:
            for cocon in contrib.cocontributors():
                if cocon.person not in cocontrib:
                    cocontrib.append(cocon.person)
                    counts.append(1)
                else:
                    idx = cocontrib.index(cocon.person)
                    counts[idx] += 1
        return zip(cocontrib, counts)

    def __str__(self):
        return ''.join(['Person: ', ', '.join(self.name), '; Items: ',
                        str(len(self.contributions))])

    def __lt__(self, other):
        return ' '.join(self.ascii_name) < ' '.join(other.ascii_name)


class Contributor(object):
    """
    Contributor
    """
    def __init__(self, item, person, name):
        self.item = weakref.proxy(item)
        self._person = weakref.proxy(person)
        self._name = name

    @property
    def ascii_name(self):
        return name_to_ascii(self._name)

    @property
    def name(self):
        return self._name

    @name.setter
    def name(self, value):
        idx = self.item.contributors.index(self)
        if self._name != value:
            self.item.update_contributor(idx, value)
            self._name = value

    @property
    def person(self):
        return self._person

    @person.setter
    def person(self, value):
        value.contributions.add(self)
        self._person.contributions.remove(self)
        self._person = weakref.proxy(value)
        self.name = value.name

    def cocontributors(self):
        return [con for con in self.item.contributors if con is not self]

    def __str__(self):
        return ''.join(['Contrib: ', ', '.join(self.name)])


bibclean/peoplenames.py
from unidecode import unidecode
import regex


reBRK = regex.compile(r"\b((\p{Ll})\p{Ll}*)\b|" +
                      r"\b((Ó)\s?\p{Lu}\p{Ll}*)|" +
                      r"((\p{Lu})('\p{Lu})?\p{Ll}*)")


def break_name(name):
    parts = [[], []]
    for idx, nm in enumerate(name):
        for p in reBRK.findall(nm):
            if p[0] != '':
                parts[idx].append(p[:2])
            elif p[2] != '':
                parts[idx].append((p[2], p[3]))
            else:
                parts[idx].append((p[4], p[5]))
    return parts


def part_comp(parts, comp):
    return [p[comp] for p in parts]


def inv_idx(idxs, N):
    """
    numbers between in range(0, N) that are not in idxs
    """
    return [idx for idx in range(0, N) if idx not in idxs]


def name_comp(a, b):
    """
    Calculate comparison score between two Unicode name tuples (last, first)
    """
    # Break names into parts --------------------------------------------------
    # Name a
    ap = break_name(a)           # Unicode parts
    aa = name_to_ascii(a)        # ASCII name
    aap = break_name(aa)         # ASCII parts
    lnap = part_comp(aap[0], 0)  # Full ASCII last name parts
    # Name b
    bp = break_name(b)
    ba = name_to_ascii(b)
    bap = break_name(ba)
    lnbp = part_comp(bap[0], 0)
    # Score Last Name ---------------------------------------------------------
    # Get array of tuples [(ai,bi)] where lnap[ai] == lnbp[bi]
    lnm = [(idx, lnbp.index(n)) for idx, n in enumerate(lnap) if n in lnbp]
    if not lnm:  # No matches for last name
        return 0
    score = 0.0
    if a[0] == b[0]:                               # Last name Uni full match
        score += 2
    elif aa[0] == ba[0]:                           # Last name ASCII full match
        score += 1.75
    elif len(lnm) == max([len(lnap), len(lnbp)]):  # Out of order ASCII match
        for idx in lnm:
            if ap[0][idx[0]][0] == bp[0][idx[1]][0]:
                score += 0.75
            else:
                score += 0.5
    else:  # Partial match, add non-matching parts to first name parts
        for idx in lnm:
            if ap[0][idx[0]][0] == bp[0][idx[1]][0]:
                score += 0.75
            else:
                score += 0.5
        aiidx = inv_idx(part_comp(lnm, 0), len(ap[0]))
        biidx = inv_idx(part_comp(lnm, 1), len(bp[0]))
        ap[1] += [ap[0][idx] for idx in aiidx]
        aap[1] += [aap[0][idx] for idx in aiidx]
        bp[1] += [bp[0][idx] for idx in biidx]
        bap[1] += [bap[0][idx] for idx in biidx]
    # Insure that more first name parts are in  a than b
    if len(ap[1]) < len(bp[1]):
        tmp = ap
        ap = bp
        bp = tmp
        tmp = aap
        aap = bap
        bap = aap
    # Build arrays needed for scoring first names
    # Arrays for a
    afns = part_comp(ap[1], 0)    # Full Uni first name parts
    ains = part_comp(ap[1], 1)    # First name Uni initials
    aafns = part_comp(aap[1], 0)  # Full ASCII fist name parts
    aains = part_comp(aap[1], 1)  # First name ASCII initials
    # Arrays for b
    bfns = part_comp(bp[1], 0)
    bins = part_comp(bp[1], 1)
    bafns = part_comp(bap[1], 0)
    bains = part_comp(bap[1], 1)
    # Score First Name --------------------------------------------------------
    for idx in range(0, len(afns)):
        if afns[idx] in bfns:      # Full Uni
            tidx = bfns.index(afns[idx])
            score += 2/(idx+tidx+2)
        elif aafns[idx] in bafns:  # Full ASCII
            tidx = bafns.index(aafns[idx])
            score += 1.75/(idx+tidx+2)
        elif ains[idx] in bins:    # Uni Initials
            tidx = bins.index(ains[idx])
            score += 1/(idx+tidx+2)
        elif aains[idx] in bains:  # ASCII Initials
            tidx = bains.index(aains[idx])
            score += 1/(idx+tidx+2)
    return score


def name_to_ascii(name):
    """
    Changes names to pseudo-ASCII equivilent
    e.g. ('Ó Súilleabháin', 'Jörg Tanjō') -> ("O'Suilleabhain", 'Jorg Tanjo')
    """
    name = [regex.sub(r'\bÓ\s?(\p{Lu})', 'O\'\g<1>', n) for n in name]
    return tuple([regex.sub(r'`', '\'', unidecode(n)) for n in name])
